### PR TITLE
Fix NPE in `TomcatService` when the server is shutdown.

### LINKED ISSUE #3136: NPE in TomcatService when the Server stops
```java
java.lang.NullPointerException: null
	at org.apache.catalina.connector.CoyoteAdapter.postParseRequest(CoyoteAdapter.java:696)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:337)
	at com.linecorp.armeria.server.tomcat.TomcatService.lambda$serve$0(TomcatService.java:376)
	at com.linecorp.armeria.common.RequestContext.lambda$makeContextAware$3(RequestContext.java:479)
	at io.micrometer.core.instrument.internal.TimedRunnable.run(TimedRunnable.java:44)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
NPE is raised in `TomcatService` when the server is shutdown.

Modification:
- Respond with 503 if the `Connector` is in the stopped state in `TomcatService`.

Result:
- You no longer see NPE in `TomcatService` when stopping the server.
- Close #3136

### CODE REVIEW

**Comment by minwoox** [comment]
cc @policeman-kh @ghkim3221

**Review by ikhoon — APPROVED** [review]
Thanks! @minwoox

File: tomcat9/src/main/java/com/linecorp/armeria/server/tomcat/TomcatService.java [hunk 1]
```diff
@@ -400,6 +410,13 @@ public final HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) thro
         return res;
     }
 
+    private static boolean isConnectorStopped(LifecycleState connectorState) {
+        return connectorState == LifecycleState.STOPPED ||
+               connectorState == LifecycleState.DESTROYING ||
+               connectorState == LifecycleState.DESTROYED ||
+               connectorState == LifecycleState.FAILED;
```
Comment (trustin) [inline tomcat9/src/main/java/com/linecorp/armeria/server/tomcat/TomcatService.java:417]: nit: Could use `switch-case`?

Comment (minwoox) [reply tomcat9/src/main/java/com/linecorp/armeria/server/tomcat/TomcatService.java:417]: Fixed. 😉

**Comment by kojilin** [comment]
Just curious if there is the possibility that the ongoing request still passes the check and failed inside the tomcat connector's code? (Maybe rare?)

Looks like spring+tomcat does some special treatment that lets tomcat inside spring can shutdown before bean destroy by using `SmartLifecycle`? Not sure if we can use the `SmartLifecycle` feature from Spring.

**Comment by minwoox** [comment]
> Looks like spring+tomcat does some special treatment that lets tomcat inside spring can shutdown before bean destroy by using SmartLifecycle? Not sure if we can use the SmartLifecycle feature from Spring.

I think we cannot use any features from Spring because armeria-tomcat does not depend on Spring.

> Just curious if there is the possibility that the ongoing request still passes the check and failed inside the tomcat connector's code? (Maybe rare?)

Yeah, I worried about it. But the service becomes `null` after `LifecycleState.DESTOYING` is set which is set after `LifecycleState.STOPPED`. So we have more times to prevent the situation.
I have tested in my local and I couldn't reproduce.
But if you are still worried about it, I will make a wrapper that has the not nullable service. 😉

**Comment by kojilin** [comment]
> Yeah, I worried about it. But the service becomes null after LifecycleState.DESTOYING is set which is set after LifecycleState.STOPPED. So we have more times to prevent the situation.
I have tested in my local and I couldn't reproduce.
But if you are still worried about it, I will make a wrapper for the connector that has the not nullable service. 😉

I think PR is no problem. We can check if it happens in the future.

> I think we cannot use any features from Spring because armeria-tomcat does not depend on Spring.

My original thought is below in `AbstractArmeriaAutoConfiguration`, but not sure if it's preferred by the spring framework and we can enhance it in another PR. 
(Not familiar with `SmartLifecycle`..).
```
public SmartLifecycle armeriaServer(...){
        ...
        return new SmartLifecycle() {
            @Override
            public void start() {
                // We start by ourselves in bean creation.
            }

            @Override
            public void stop() {
                server.close();
            }

            @Override
            public boolean isRunning() {
                return !server.isClosed() && !server.isClosing();
            }

            @Override
            public boolean isAutoStartup() {
                // We start by ourselves in bean creation.
                return false;
            }
        };
}
```

**Comment by minwoox** [comment]
This `TomcatService` should not use Spring so we might need a wrapper to support `SmartLifecycle` in tomcat module.

> Just curious if there is the possibility that the ongoing request still passes the check and failed inside the tomcat connector's code? (Maybe rare?)

I'm a bit wondering how the `LifeCycle` prevents this situation. 🤔

**Comment by kojilin** [comment]
I didn't describe clearly, my original idea is to make ArmeriaServer start the shutdown process(Enable graceful shutdown) before tomcat. This can also avoid 503?

If we implement SmartLifeCycle, it will trigger before the usual bean close.
https://github.com/spring-projects/spring-framework/blob/master/spring-context/src/main/java/org/springframework/context/support/AbstractApplicationContext.java#L1070-L1080

And WebServerStartStopLifecycle in spring-boot also wraps web server that makes tomcat bean stop first I think. And  priority of `WebServerStartStopLifecycle#getPhase` is default value -1.

**Review by trustin — APPROVED** [review]

**Comment by trustin** [comment]
@kojilin @minwoox Shall we create a new issue for that? :bow:

**Comment by kojilin** [comment]
Let me create the PR and also paste the test maybe.^^

**Comment by minwoox** [comment]
Thanks @kojilin!

**Pull request merged**